### DISCUSSION TITLE
performance: revalidate, copy

### DISCUSSION THREAD

**Discussion by dfroger** [pydantic/pydantic#6967] [Question]
Hi,

I'm having performance issues with code using Pydantic 1.9.1. For example, when [profiing](https://docs.python.org/3/library/profile.html) a script, I see that 98.5% of the time in spent in copy, that I solved with some [construct](https://docs.pydantic.dev/1.10/usage/models/#creating-models-without-validation).

I plan to migrate to Pydantic 2 soon, but it's not clear for me how to reason on revalidation and copy in V2 (not sure if there are changes between V1 and V2 on this subject?). My code manipulates large list of Pydantic models, and build Pydantic model from other. This can be illustrated with code like this:

```python
from typing import List

from pydantic import BaseModel, model_validator


class Foo(BaseModel, revalidate_instances="never"):
    x: int

    @model_validator(mode="after")
    def log_validation(self) -> "Foo":
        print(f"validated Foo: {self.x}")
        return self


class Bar(BaseModel, revalidate_instances="never"):
    y: int
    foos: List[Foo]

    @model_validator(mode="after")
    def log_validation(self) -> "Bar":
        print(f"validated Bar: {self.y}")
        return self


class Baz(BaseModel, revalidate_instances="never"):
    z: int
    bar: Bar

    @model_validator(mode="after")
    def log_validation(self) -> "Baz":
        print(f"validated Baz: {self.z}")
        return self


class Bam(BaseModel, revalidate_instances="never"):
    baz: Baz


foo1 = Foo(x=1)
foo2 = Foo(x=2)
bar = Bar(y=33, foos=[foo1, foo2])
baz = Baz(z=444, bar=bar)
bam = Bam(baz=baz)

if id(bam.baz.bar.foos[0]) == id(foo1):
    print("Seems there is no copy")
else:
    print("Seems there are no copies")
```

```
validated Foo: 1
validated Foo: 2
validated Foo: 1
validated Foo: 2
validated Bar: 33
validated Bar: 33
validated Baz: 444
validated Baz: 444
Seems there is no copy
```

In such code, I would like for example:
- when creating `bar`: ensure that `foos` is a list of `Foo`, but don't revalidate/copy each item
- when creating `baz`: ensure that `bar` is a `Bar`, but don't revalidate/copy it.
- etc

How who should I reason on revalidate/copy, and write code to avoid them? For example, in this situation:

```python
from pydantic import BaseModel, model_validator


class Foo(BaseModel, revalidate_instances="never"):
    x: int

    @model_validator(mode="after")
    def log_validation(self) -> "Foo":
        print(f"validated Foo: {self.x}")
        return self


class Bar(BaseModel, revalidate_instances="never"):
    foo: Foo


foo = Foo(x=1)
bar = Bar(foo=foo)
```

```python
validated Foo: 1
validated Foo: 1
```

Why is `foo` validated twice? I would expect `revalidate_instances="never"` to prevent this.

Thanks!

**Comment by Atheuz** [comment]
I have the same problem of being confused about why it is re-validating things, but you mentioned `construct` and there's a similar method in pydantic v2 called `model_construct` that maybe you could use.

```
from pydantic import BaseModel, model_validator

class Foo(BaseModel):
    x: int
    @model_validator(mode="after")
    def log_validation(self) -> "Foo":
        print(f"validated Foo: {self.x}")
        return self

class Bar(BaseModel):
    foo: Foo
```

```
foo = Foo(x=1)
validated Foo: 1
```

```
bar = Bar.model_construct(foo=foo)
# ...nothing...
```

I would rather have that `model_validator` works correctly, and I'm probably not gonna use this, but I wanted to highlight it for you since you mention you used `construct` previously.
